server/src/main/java/org/elasticsearch/gateway/MetaStateService.java

```diff
@@ -172,20 +173,32 @@ List<IndexMetadata> loadIndicesStates(Predicate<String> excludeIndexPathIdsPredi
         for (String indexFolderName : nodeEnv.availableIndexFolders(excludeIndexPathIdsPredicate)) {
             assert excludeIndexPathIdsPredicate.test(indexFolderName) == false
                 : "unexpected folder " + indexFolderName + " which should have been excluded";
-            IndexMetadata indexMetadata = IndexMetadata.FORMAT.loadLatestState(
-                logger,
-                namedXContentRegistry,
-                nodeEnv.resolveIndexFolder(indexFolderName)
-            );
-            if (indexMetadata != null) {
-                final String indexPathId = indexMetadata.getIndex().getUUID();
-                if (indexFolderName.equals(indexPathId)) {
-                    indexMetadataList.add(indexMetadata);
+            try {
+                IndexMetadata indexMetadata = IndexMetadata.FORMAT.loadLatestState(
+                    logger,
+                    namedXContentRegistry,
+                    nodeEnv.resolveIndexFolder(indexFolderName)
+                );
+                if (indexMetadata != null) {
+                    final String indexPathId = indexMetadata.getIndex().getUUID();
+                    if (indexFolderName.equals(indexPathId)) {
+                        indexMetadataList.add(indexMetadata);
+                    } else {
+                        throw new IllegalStateException(
+                            "[" + indexFolderName + "] invalid index folder name, rename to [" + indexPathId + "]"
+                        );
+                    }
                 } else {
-                    throw new IllegalStateException("[" + indexFolderName + "] invalid index folder name, rename to [" + indexPathId + "]");
+                    logger.debug("[{}] failed to find metadata for existing index location", indexFolderName);
                 }
-            } else {
-                logger.debug("[{}] failed to find metadata for existing index location", indexFolderName);
+            } catch (IOException e) {
+                // index directory was deleted concurrently, ignoring
+            } catch (ElasticsearchException e) {
+                if ((e.getCause() instanceof IOException) == false) {
+                    throw e;
+                }
+            } catch (IllegalStateException e) {
+                // partially deleted index, ignoring
             }
```

Comment:
I am not very happy on how this looks.

I wonder if we should instead try to find locks with the same index UUID and acquire if they are present?

Reply:
I don't think that'll solve the problem, the lock might not exist when you check, but then come back into existence to delete the data shortly after